File: Week_6_Python/credit.py
# Luhn's Algorith to check credit card number

# num = 4003600000000014

##### sum of digits #####
def digitsSum(n):
    n_string = str(n)
    c = 0
    for i in range(len(n_string)):
        c += int(n_string[i])
    return c

###### check validity of card number #####


def validateCard(card_num):
    ## card num as string ##
    num_string = str(card_num)
    l = len(num_string)
    sw = int(num_string[:2])
    count1, count2 = 0, 0
    if l == 14 or l < 13 or l > 16:
        print("INVALID")
        return
    for i in range(- l, 0):
        ## every other digits ##
        if i % 2 == 0:
            x = int(num_string[i]) * 2
            count1 += digitsSum(x)
        ## rest of the digits ##
        else:
            count2 += int(num_string[i])

    total = count1 + count2
    if total % 10 == 0:
        if l == 15 and sw in [34, 37]:
            print("AMEX")
        elif int(sw / 10) == 4 and l in [13, 16]:
            print("VISA")
        elif 50 < sw < 56 and l == 16:
            print("MASTERCARD")
        else:
            print("INVALID")
    else:
        print("INVALID")


num = int(input("Number: "))
validateCard(num)

File: Week_1_C/credit.c
#include <stdio.h>
#include <cs50.h>

// main function
int main(void)
{
    int count();
    void credit();
    // printf("Number : ");
    long int card_number = get_long("Number : ");
    credit(card_number);
}

// count sum of digits
int count(int num)
{
    int sum = 0;
    while (num > 0)
    {
        sum = sum + num % 10;
        num = num / 10;
    }
    return sum;
}

// count number of digits
int length(long int num)
{
    int len = 0;
    while (num > 0)
    {
        len++;
        num = num / 10;
    }
    return len;
}

// return first two digits
int starts_with(long int num)
{
    while (num > 99)
    {
        num = num / 10;
    }
    return num;
}
// check if card number is valid or not
void credit(long int num)
{
    int l = length(num);
    int sw = starts_with(num);
    int i = 1;
    int sum = 0;
    int sum2 = 0;
    if (l == 14 || l < 13 || l > 16)
    {
        printf("INVALID\n");
        return;
    }
    while (num > 0)
    {
        if (i % 2 == 0)
        {
            sum = sum + count(2 * (num % 10));
        }
        else
        {
            sum2 = sum2 + num % 10;
        }
        num = num / 10;
        i++;
    }
    if ((sum + sum2) % 10 == 0)
    {
        if (l == 15 && (sw == 34 || sw == 37))
        {
            printf("AMEX\n");         
        }
        else if (sw / 10 == 4 && (l == 13 || l == 16))
        {
            printf("VISA\n");     
        }
        else if (sw > 50 && sw < 56 && l == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
